docs: add documentation for discovery chains (#6293)
hashicorp/consul: website/source/docs/internals/discovery-chain.html.md

---
layout: "docs"
page_title: "Discovery Chain"
sidebar_current: "docs-internals-discovery-chain"
description: |-
  The service discovery process can be modeled as a "discovery chain" which passes through three distinct stages: routing, splitting, and resolution. Each of these stages is controlled by a set of configuration entries.
---

> # Discovery Chain <sup>beta</sup>

Reviewer: if this ends up not being merged until GA we should make a note to fix this as we may have already made a 'beta' removal pass in the docs. Hopefully can be merged before though.

Author: We have not made a beta pass yet.

Author: Made an issue so we don't forget: https://github.com/hashicorp/consul/issues/6349

-> We recommended reading the high-level documentation for [L7 Traffic
Management](/docs/connect/l7-traffic-management.html) first. This page
discusses lower-level aspects that make up the implementation of that feature
set.

The service discovery process can be modeled as a "discovery chain" which
passes through three distinct stages: routing, splitting, and resolution. Each
of these stages is controlled by a set of [configuration
entries](/docs/agent/config_entries.html).

> -> **NOTE** The discovery chain is only used to discover

Reviewer: should we add "currently" to this? It's at least somewhat intended to use this for HTTP/DNS eventually... OK with leaving it though given how non-concrete that timeline is...

Reviewer 2: I would added "currently" or try to soften the language. For example: "The discover chain enables users to discover proxy upstream services so that......." I'd also add this to the previous paragraph.

[Connect](/docs/connect/index.html) proxy upstreams.

## Configuration

The configuration entries used in the discovery chain are designed to be simple
to read and modify for narrowly tailored changes, but at discovery-time the
various configuration entries interact in more complex ways. For example:

* If a [`service-resolver`](/docs/agent/config-entries/service-resolver.html)
  is created with a [service
  redirect](/docs/agent/config-entries/service-resolver.html#service) defined,
  then all references made to the original service in any other configuration
  entry is replaced with the redirect destination.

* If a [`service-resolver`](/docs/agent/config-entries/service-resolver.html)
  is created with a [default
  subset](/docs/agent/config-entries/service-resolver.html#defaultsubset)
  defined then all references made to the original service in any other
  configuration entry that did not specify a subset will be replaced with the
  default.

* If a [`service-splitter`](/docs/agent/config-entries/service-splitter.html)
  is created with a [service

>   split](/docs/agent/config-entries/service-splitter.html#splits) that has its

Reviewer: ```suggestion
  split](/docs/agent/config-entries/service-splitter.html#splits), and the target service has its
```

  aggregate traffic split is ultimately configured in the proxy.

* [`service-resolver`](/docs/agent/config-entries/service-resolver.html)
  redirect loops must be rejected as invalid.

* [`service-router`](/docs/agent/config-entries/service-router.html) and
  [`service-splitter`](/docs/agent/config-entries/service-splitter.html)
  configuration entries require an L7 compatible protocol be set for the
  service via either a
  [`service-defaults`](/docs/agent/config-entries/service-defaults.html) or
  [`proxy-defaults`](/docs/agent/config-entries/proxy-defaults.html) config
  entry. Violations must be rejected as invalid.

* If an [upstream
  configuration](/docs/connect/registration/service-registration.html#upstream-configuration-reference)
  [`datacenter`](/docs/connect/registration/service-registration.html#datacenter)
  parameter is defined then any configuration entry that does not explicitly
  refer to a desired datacenter should use that value from the upstream.

## Compilation

To correctly interpret a collection of configuration entries as a valid
discovery chain we first compile them into a form more directly usable by the

> layers responsible for configuration connect sidecar proxies.

Reviewer: ```suggestion
layers responsible for configuring Connect sidecar proxies.
```

You can interact with the compiler directly using the [discovery-chain
API](/api/discovery-chain.html).

### Compilation Parameters

* **Service Name** - The service being discovered by name.
* **Datacenter** - The datacenter to use as the basis of compilation. 
* **Overrides** - Discovery-time tweaks to apply when compiling. These should
  be derived from either the
  [proxy](/docs/connect/registration/service-registration.html#complete-configuration-example)
  or
  [upstream](/docs/connect/registration/service-registration.html#upstream-configuration-reference)
  configurations if either are set.

### Compilation Results

The response is a single wrapped `CompiledDiscoveryChain` field:

```json
{
    "Chain": {...<CompiledDiscoveryChain>...}
}
```

#### `CompiledDiscoveryChain`

The chain encodes a digraph of [nodes](#discoverygraphnode) and
[targets](#discoverytarget). Nodes are the compiled representation of various
discovery chain stages and targets are instructions on how to use the [health
API](/api/health.html#list-nodes-for-connect-capable-service) to retrieve
relevant service instance lists.

> You should traverse the nodes starting with [`StartNode`](#startnode). The

Reviewer: I don't see any anchor or heading matching `#startnode`. Same for others below - did you intent to add those around the parameters below?

Author: Lists with the leading term wrapped by backticks (like the consul agent options page) will link each term.

#### `DiscoveryGraphNode`

A single node in the compiled discovery chain.

- `Type` `(string)` - The type of the node. Valid values are: `router`,
  `splitter`, and `resolver`.

- `Name` `(string)` - The unique name of the node.

- `Routes` `(array<DiscoveryRoute>)` - Only set for `Type:router`. List of routes to
  render.

  - `Definition` `(ServiceRoute)` - Relevant portion of underlying
    `service-router`
    [route](/docs/agent/config-entries/service-router.html#routes).

  - `NextNode` `(string)` - The name of the next node in the chain in [`Nodes`](#nodes).

- `Splits` `(array<DiscoverySplit>)` - Only set for `Type:splitter`. List of traffic
  splits.

  - `Weight` `(float32)` - Copy of underlying `service-splitter`
    [`weight`](/docs/agent/config-entries/service-splitter.html#weight) field.

  - `NextNode` `(string)` - The name of the next node in the chain in [`Nodes`](#nodes).

- `Resolver` `(DiscoveryResolver: <optional>)` - Only set for `Type:resolver`. How
  to resolve the service instances.

  - `Default` `(bool)` - Set to true if no `service-resolver` config entry is
    defined for this node and the default was synthesized.

>   - `ConnectTimeout` `(duration)` - Copy of the underlying `service-resolver`

Reviewer: It's actually an integer in nanoseconds it seems. I'd expect `(duration)` to be a string like `5s`

---
layout: "docs"
page_title: "Discovery Chain"
sidebar_current: "docs-internals-discovery-chain"
description: |-
  The service discovery process can be modeled as a "discovery chain" which passes through three distinct stages: routing, splitting, and resolution. Each of these stages is controlled by a set of configuration entries.
---

# Discovery Chain <sup>beta</sup>

> ~> This topic is part of a [low-level API](/api/discovery-chain.html)

Reviewer 2: This intro is great!

  own `service-splitter` then the overall effect is flattened and only a single 
  aggregate traffic split is ultimately configured in the proxy.

* [`service-resolver`](/docs/agent/config-entries/service-resolver.html)
  redirect loops must be rejected as invalid.

* [`service-router`](/docs/agent/config-entries/service-router.html) and
  [`service-splitter`](/docs/agent/config-entries/service-splitter.html)
  configuration entries require an L7 compatible protocol be set for the
  service via either a
  [`service-defaults`](/docs/agent/config-entries/service-defaults.html) or
  [`proxy-defaults`](/docs/agent/config-entries/proxy-defaults.html) config
  entry. Violations must be rejected as invalid.

* If an [upstream
  configuration](/docs/connect/registration/service-registration.html#upstream-configuration-reference)
  [`datacenter`](/docs/connect/registration/service-registration.html#datacenter)
  parameter is defined then any configuration entry that does not explicitly
  refer to a desired datacenter should use that value from the upstream.

## Compilation

To correctly interpret a collection of configuration entries as a valid

> discovery chain we first compile them into a form more directly usable by the

Reviewer 2: ```suggestion
discovery chain, we first compile them into a form more directly usable by the
```
